graph_worker.cpp

```diff
@@ -19,37 +29,74 @@ void GraphWorker::startWorkers(Graph *_graph, BufferTree *_bf) {
 	}
 }
 
-void GraphWorker::stopWorkers() {
+void GraphWorker::stop_workers() {
 	shutdown = true;
+	workers[0]->bf->bypass_wait(true); // make the GraphWorkers bypass waiting in queue
+	pause_condition.notify_all();      // tell any paused threads to continue and exit
 	for (int i = 0; i < num_groups; i++) {
 		delete workers[i];
 	}
 	delete workers;
 }
 
+void GraphWorker::pause_workers() {
+	paused = true;
+	workers[0]->bf->bypass_wait(true); // make the GraphWorkers bypass waiting in queue
+
+	// wait until all GraphWorkers are paused
+	std::unique_lock<std::mutex> lk(pause_lock);
+	pause_condition.wait(lk, []{
+		for (int i = 0; i < num_groups; i++)
+			if (!workers[i]->get_thr_paused()) return false;
+		return true;
+	});
+	lk.unlock();
+}
+
+void GraphWorker::unpause_workers() {
+	workers[0]->bf->bypass_wait(false); // buffer-tree operations should block when necessary
+	paused = false;
+	pause_condition.notify_all();       // tell all paused workers to get back to work
+}
+
+/***********************************************
+ ************** GraphWorker class **************
+ ***********************************************/
 GraphWorker::GraphWorker(int _id, Graph *_graph, BufferTree *_bf) :
-  id(_id), graph(_graph), bf(_bf), thr(startWorker, this) {
+  id(_id), graph(_graph), bf(_bf), thr(start_worker, this) {
 }
 
 GraphWorker::~GraphWorker() {
-	if(id == 0)
-		bf->bypass_wait(); // to avoid race condition on shutdown. Tell threads to recheck wait
-	
-	// when a graph worker is deleted we must wait for the associated
-	// thread to finish its work. See stopWorkers and doWork
+	// join the GraphWorker thread to reclaim resources
 	thr.join();
 }
 
-void GraphWorker::doWork() {
+void GraphWorker::do_work() {
 	data_ret_t data;
 	while(true) {
-		// call get_data which will handle waiting on the queue
-		// and will enforce locking.
-		bool valid = bf->get_data(data);
-
-		if (valid)
-			graph->batch_update(data.first, data.second);
-		else if(shutdown)
+		if(shutdown)
 			return;
+		pause_lock.lock();
+		thr_paused = true; // this thread is currently paused
+		pause_lock.unlock();
+		pause_condition.notify_one(); // notify pause_workers()
+
+		// wait until we are unpaused
+		std::unique_lock<std::mutex> lk(pause_lock);
+		pause_condition.wait(lk, []{return !paused || shutdown;});
+		thr_paused = false; // no longer paused
+		lk.unlock();
+		while(!thr_paused) {
+			// call get_data which will handle waiting on the queue
+			// and will enforce locking.
+			bool valid = bf->get_data(data);
+
+			if (valid)
+				graph->batch_update(data.first, data.second);
+			else if(shutdown)
+				return;
+			else if(paused)
```

Comment:
are there race conditions or weird interactions if we don't synchronize reads of these variables?

Reply:
Yes but I don't believe it will cause problems.

Imagine we set paused to true as a worker is going through this if statement. It might be the case that the worker does not read paused==true so repeats. However, we will immediately exit from the call to bf->get_data (as the buffer tree operations are non-blocking once pause_workers is called) and check the if statement again.

Reply:
ok